pkg/server/server.go

```diff
@@ -476,3 +646,264 @@ func adaptContext(parent genericapiserver.PostStartHookContext) context.Context
+		// (across logical clusters) if the CRD of the related API Resource hasn't been added in the admin logical cluster first.
+		// The fix in this HACK is limited since the request will fail if 2 logical clusters contain CRDs for the same GVK
+		// with non-equal specs (especially non-equal schemas).
+		var crds []*apiextensionsv1.CustomResourceDefinition
+		crds, err = c.crdLister.List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+		var equal bool // true if all the found CRDs have the same spec
+		crd, equal = findCRD(name, crds)
+		if !equal {
+			err = apierrors.NewInternalError(fmt.Errorf("error resolving resource: cannot watch across logical clusters for a resource type with several distinct schemas"))
+			return nil, err
+		}
+
+		if crd == nil {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
+		}
+
+		return crd, nil
+	}
+
+	crdKey := clusters.ToClusterAwareKey(cluster.Name, name)
+	crd, err = c.crdLister.Get(crdKey)
+	if err != nil && !apierrors.IsNotFound(err) {
+		// something went wrong w/the lister
+		return nil, err
+	}
+
+	if crd != nil {
+		return crd, nil
+	}
+
+	// Check for API inheritance
+	const adminClusterName = "admin" // FIXME
+	targetWorkspaceKey := clusters.ToClusterAwareKey(adminClusterName, cluster.Name)
+	workspace, err := c.workspaceLister.Get(targetWorkspaceKey)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
+		}
+
+		return nil, err
+	}
+
+	if workspace.Spec.InheritFrom == "" {
+		return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
+	}
+
+	sourceWorkspaceKey := clusters.ToClusterAwareKey(adminClusterName, workspace.Spec.InheritFrom)
+	if _, err := c.workspaceLister.Get(sourceWorkspaceKey); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
+		}
+
+		return nil, err
+	}
+
+	sourceWorkspaceCRDKey := clusters.ToClusterAwareKey(workspace.Spec.InheritFrom, name)
+	crd, err = c.crdLister.Get(sourceWorkspaceCRDKey)
+	return crd, err
+}
+
+func findCRD(crdName string, crds []*apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, bool) {
+	var crd *apiextensionsv1.CustomResourceDefinition
+
+	for _, aCRD := range crds {
+		if aCRD.Name != crdName {
+			continue
+		}
+		if crd == nil {
+			crd = aCRD
+		} else {
+			if !equality.Semantic.DeepEqual(crd.Spec, aCRD.Spec) {
+				return crd, false
+			}
+		}
+	}
+
+	return crd, true
+}
+
+type kcpAPIExtensionsSharedInformerFactory struct {
```

Comment:
I should probably comment on this chain of structs - it exists solely to be able to override the CRD lister you get from the CRD informer with our own.